src/hotspot/share/memory/heapInspection.cpp

```diff
@@ -217,9 +217,10 @@ bool KlassInfoTable::record_instance(const oop obj) {
   // elt may be null if it's a new klass for which we
   // could not allocate space for a new entry in the hashtable.
   if (elt != nullptr) {
-    elt->set_count(elt->count() + 1);
-    elt->set_words(elt->words() + obj->size());
-    _size_of_instances_in_words += obj->size();
+    elt->atomic_inc_count();
```

Comment:
We don't want everyone to be penalized for the atomic operations even when they are executing this method single-threaded. I'd use a new method that is MT-safe and uses atomic operations, and use those for the multithreaded case.

See also my other comment about a future refactor where we do away with atomic ops and use thread-local tables with a final merge step.